File: C#/RotatingRobot.cs
/* RotatingRobot.cs
* Don-Thuan Le
* CPSC3200
* Date: 5/8/2020
* version: 1.0
* 
* Class Invariants:
* When instantiated, 3 sensors are made (up, left and right), 1 actuator is made(up), the starting position is 
* 5,5, the grid is loaded in from the command line and the robot direction is Up.
* 
* Interface Invariants:
* RotatingRobot cannot rotate 180 degrees, can only turn 90 degrees to the left or 90 degrees to the right. Move one
* either rotates the robot or moves the robot. Robot will never go back the way it came from, it will stop at the edge
* of the maze.
*/


using System;
using System.Collections.Generic;
using System.Text;

namespace P5
{
    public class RotatingRobot : Robot
    {
        private string Next;
        protected Direction RobotDirection;

        public RotatingRobot(string filePath) : base(filePath)
        {
            RobotDirection = Direction.Up;
            Sensors[Direction.Left] = new Sensor(Direction.Left);
            Sensors[Direction.Right] = new Sensor(Direction.Right);
        }

        public RotatingRobot(string filePath, int posR1, int posC1) : base(filePath)
        {
            PosR = posR1;
            PosC = posC1;
            RobotDirection = Direction.Up;
            Sensors[Direction.Left] = new Sensor(Direction.Left);
            Sensors[Direction.Right] = new Sensor(Direction.Right);
        }

        public override bool isValid()
        {
            if (IsPoweredup())
            {
                foreach (var item in Sensors)
                {
                    if (item.Value.IsValid(MainGrid, PosR, PosC))
                    {
                        Next = NextDirection(item.Key);
                        return true;
                    }
                }
            }
            return false;
        }

        public string NextDirection(Direction directionTo)
        {
            switch (RobotDirection)
            {
                case Direction.Up:
                    if (directionTo == Direction.Left)
                        return "left";
                    else if (directionTo == Direction.Right)
                        return "right";
                    break;
                case Direction.Down:
                    if (directionTo == Direction.Left)
                        return "right";
                    else if (directionTo == Direction.Right)
                        return "left";
                    break;
                case Direction.Left:
                    if (directionTo == Direction.Up)
                        return "right";
                    else if (directionTo == Direction.Down)
                        return "left";
                    break;
                case Direction.Right:
                    if (directionTo == Direction.Up)
                        return "left";
                    else if (directionTo == Direction.Down)
                        return "right";
                    break;
            }
            if (RobotDirection == directionTo)
                return "straight";
            return null;
        }

        public override void move()
        {
            while (moveOne())
            {
                Console.WriteLine($"rotating robot x:{this.getPosR()} y:{this.getPosC()}");
            }
        }

        public override bool moveOne()
        {
            ChargeSensors();
            if (isValid())
            {
                switch (Next)
                {
                    case "left":
                        rotate(Next);
                        break;
                    case "right":
                        rotate(Next);
                        break;
                    case "straight":
                        Actuators[RobotDirection].MoveForward(this);
                        break;
                }
                return true;
            }
            return false;
        }

        // PreConditions: Direction for robot to turn to
        // PostConditions: Robot Direction is changed respectively
        public void rotate(string turn)
        {
            turn.ToLower();

            switch (RobotDirection)
            {
                case Direction.Up:
                    if (turn == "right")
                    {
                        RobotDirection = Direction.Right;
                    }
                    else if (turn == "left")
                    {
                        RobotDirection = Direction.Left;
                    }
                    break;
                case Direction.Right:
                    if (turn == "right")
                    {
                        RobotDirection = Direction.Down;
                    }
                    else if (turn == "left")
                    {
                        RobotDirection = Direction.Up;
                    }
                    break;
                case Direction.Down:
                    if (turn == "right")
                    {
                        RobotDirection = Direction.Left;
                    }
                    else if (turn == "left")
                    {
                        RobotDirection = Direction.Right;
                    }
                    break;
                case Direction.Left:
                    if (turn == "right")
                    {
                        RobotDirection = Direction.Up;
                    }
                    else if (turn == "left")
                    {
                        RobotDirection = Direction.Down;
                    }
                    break;
            }

            var SensorsTemp = new Dictionary<Direction, Sensor>();

            foreach (var item in Sensors)
            {
                var direction = item.Value.SensorRotate(turn);
                SensorsTemp[direction] = item.Value;
            }
            Sensors = SensorsTemp;

            var ActuatorsTemp = new Dictionary<Direction, Actuator>();

            foreach (var item in Actuators)
            {
                var direction = item.Value.ActuatorRotate(turn);
                ActuatorsTemp[direction] = item.Value;
            }
            Actuators = ActuatorsTemp;
        }

        public Direction getRobotDirection()
        {
            return RobotDirection;
        }
    }
}

/*
 * Implementation Invariants:
 * 
 * RotatingRobot is a class that has inherited the Robot class. When initialized, it has 3 
 * sensors (up, left, and right) and there is one actuator still (up). Overrided isValid helps
 * identify which direction is next from the three sensors. One identified, the robot rotates 
 * to that direction and the sensors and actuator rotate accordingly to continue the maze. In the 
 * moveOne function, rotate is counted is one move.
 */

File: C++/rotatingRobot.cpp
#include "rotatingRobot.h"
#include <iostream>
#include <string>

rotatingRobot::rotatingRobot(grid& main) : robot(main) {
	RobotDirection = Up;

    list<sensor*>* sensorsLeft = new list<sensor*>();
    list<sensor*>* sensorsRight = new list<sensor*>();
	makeSensors(sensorsLeft, Left);
	makeSensors(sensorsRight, Right);

    sensors->insert({ Left, sensorsLeft });
    sensors->insert({Right, sensorsRight});
}

rotatingRobot::rotatingRobot(const rotatingRobot& obj) : robot(obj){
    RobotDirection = obj.RobotDirection;
    Next = obj.Next;
}

rotatingRobot& rotatingRobot::operator=(const rotatingRobot& obj){
    if (this != &obj){
        RobotDirection = obj.RobotDirection;
        Next = obj.Next;
        this->robot::operator=(obj);
    }
    return *this;
}

rotatingRobot::~rotatingRobot(){
}

bool rotatingRobot::isValid() {
    list<sensor*>* currRobotDirectionSensors = (*sensors)[RobotDirection];
    if(!allSensorBatteryCheck(currRobotDirectionSensors))
        chargeSensors();
    for (sensor* sensor : *currRobotDirectionSensors){
        if (sensor->isValid(mainGrid, posR, posC))
        {
            Next = nextDirection(RobotDirection);
            return true;
        }
    }

    for (std::pair<Direction, list<sensor*>*> element : *sensors) {
        list<sensor*>* currSensors = element.second;
        if(!allSensorBatteryCheck(currSensors))
            chargeSensors();
        for(sensor* temp: *currSensors){
            if (element.first != oppositeDirection(RobotDirection) && temp->isValid(mainGrid, posR, posC)){
                Next = nextDirection(element.first);
                return true;
            }
        }
    }
    return false;
}

std::string rotatingRobot::nextDirection(Direction directionTo){
    switch (RobotDirection){
        case Up:
            if (directionTo == Left)
                return "left";
            else if (directionTo == Right)
                return "right";
            break;
        case Down:
            if (directionTo == Left)
                return "right";
            else if (directionTo == Right)
                return "left";
            break;
        case Left:
            if (directionTo == Up)
                return "right";
            else if (directionTo == Down)
                return "left";
            break;
        case Right:
            if (directionTo == Up)
                return "left";
            else if (directionTo == Down)
                return "right";
            break;
    }
    if (RobotDirection == directionTo)
        return "straight";
    return NULL;
}

bool rotatingRobot::moveOne(){
    if(isValid()){
        if(Next == "left")
            rotate(Next);
        else if(Next == "right")
            rotate(Next);
        else if(Next == "straight"){
            (*actuators)[RobotDirection]->moveForward(*this);
            CountMoves++;
        }
        return true;
    }
    return false;
}

void rotatingRobot::rotate(std::string turn){
    switch (RobotDirection)
    {
        case Up:
            if (turn == "right")
            {
                RobotDirection = Right;
            }
            else if (turn == "left")
            {
                RobotDirection = Left;
            }
            break;
        case Right:
            if (turn == "right")
            {
                RobotDirection = Down;
            }
            else if (turn == "left")
            {
                RobotDirection = Up;
            }
            break;
        case Down:
            if (turn == "right")
            {
                RobotDirection = Left;
            }
            else if (turn == "left")
            {
                RobotDirection = Right;
            }
            break;
        case Left:
            if (turn == "right")
            {
                RobotDirection = Up;
            }
            else if (turn == "left")
            {
                RobotDirection = Down;
            }
            break;
    }
    
    map<Direction, list<sensor*>*>* SensorsTemp = new map<Direction, list<sensor*>*>();
    
    for (std::pair<Direction, list<sensor*>*> element : *sensors) {
        Direction newDirection;
        for(sensor* temp: *(element.second)){
            newDirection = temp->sensorRotate(turn);
        }
        (*SensorsTemp)[newDirection] = element.second;
    }
    delete sensors;
    sensors = SensorsTemp;

    map<Direction, actuator*>* ActuatorsTemp = new map<Direction, actuator*>();
    
    for (std::pair<Direction, actuator*> element : *actuators) {
        Direction newDirection = element.second->actuatorRotate(turn);
        (*ActuatorsTemp)[newDirection] = element.second;
    }
    delete actuators;
    actuators = ActuatorsTemp;
}

Direction rotatingRobot::getRobotDirection(){
    return RobotDirection;
}

Direction rotatingRobot::oppositeDirection(Direction currDirection)
{
    switch (currDirection)
    {
        case Up:
            return Down;
        case Down:
            return Up;
        case Left:
            return Right;
        case Right:
            return Left;
        default:
            break;
    }
    return None;
}




